endpoints/v2/blob.py

```diff
@@ -51,82 +56,116 @@
-    if direct_download_url:
-        logger.debug("Returning direct download URL")
-        resp = redirect(direct_download_url)
-        resp.headers.extend(headers)
-        return resp
-
-    # Close the database connection before we stream the download.
-    logger.debug("Closing database connection before streaming layer data")
-    headers.update(
-        {
+class CheckBlobExists(MethodView):
+    decorators = [
+        cache_control(max_age=31436000),
+        anon_allowed,
+        require_repo_read,
+        process_registry_jwt_auth(scopes=["pull"]),
+        parse_repository_name(),
+        disallow_for_account_recovery_mode,
+    ]
+
+    def head(self, namespace_name, repo_name, digest):
+        blob = registry_model.get_cached_repo_blob(model_cache, namespace_name, repo_name, digest)
+        if blob is None:
+            try:
+                proxy = setup_proxy(namespace_name, repo_name)
+                resp = proxy.blob_exists(digest)
+                return Response(**resp)
+            except ProxyNotSupported:
+                pass
+            raise BlobUnknown()
+
+        return self.check_exists(blob, namespace_name, digest)
+
+    def check_exists(self, blob, namespace_name, digest):
+        headers = {
+            "Docker-Content-Digest": digest,
             "Content-Length": blob.compressed_size,
             "Content-Type": BLOB_CONTENT_TYPE,
         }
-    )
 
-    with database.CloseForLongOperation(app.config):
-        # Stream the response to the client.
-        return Response(
-            storage.stream_read(blob.placements, path),
-            headers=headers,
+        # If our storage supports range requests, let the client know.
+        if storage.get_supports_resumable_downloads(blob.placements):
+            headers["Accept-Ranges"] = "bytes"
+
+        return Response(headers=headers)
+
+
+v2_bp.add_url_rule(
+    BLOB_DIGEST_ROUTE,
+    view_func=CheckBlobExists.as_view("check_blob_exists"),
+)
+
+
+class DownloadBlob(MethodView):
+    decorators = [
+        cache_control(max_age=31536000),
+        check_region_blacklisted(BlobDownloadGeoBlocked),
+        anon_allowed,
+        require_repo_read,
+        process_registry_jwt_auth(scopes=["pull"]),
+        parse_repository_name(),
+        disallow_for_account_recovery_mode,
+    ]
+
+    def get(self, namespace_name, repo_name, digest):
+        blob = registry_model.get_cached_repo_blob(model_cache, namespace_name, repo_name, digest)
+        if blob is None:
+            try:
+                proxy = setup_proxy(namespace_name, repo_name)
+                media_type = request.headers.get("Accept", None)
+                resp = proxy.get_blob(digest, media_type)
+                return Response(**resp)
+            except ProxyNotSupported:
+                pass
+            raise BlobUnknown()
+
+        return self.download_blob(blob, namespace_name, digest)
+
+    def download_blob(self, blob, namespace_name, digest):
```

Comment:
The previous `download_blob` view became a method of the `DownloadBlob` class. No changes were made to it, other than adding the blob as a method argument.